Process Manager and Seat Assignments Aggregate

AlexanderLitus opened the pull request:
Is based on PR #11.

Commit:
Use SeatAssignmentsId in messages.

Commit:
Merge branch 'master' into seat-assignments-aggregate

Commit:
Add code style settings.

Commit:
Update inspection profile.

Commit:
Update Spine dependency.

Commit:
Replace package imports.

Commit:
Add /*package*/ comments to package members.

Commit:
Remove redundant warning suppression.

Commit:
Change style settings.

Commit:
Add seat assignments aggregate handling `assign seat` command.

Commit:
Remove redundant warning suppression.

Commit:
Handle UnassignSeat command.

Commit:
Post commands and also post generated events.

Commit:
Fix layout.

Commit:
Handle and apply all commands remained in Seats Assignments aggregate.
Move SeatAssignment definition to use it in events/commands.
Add seats field to OrderConfirmed event.

Commit:
Make fields private.

Commit:
Move and create utility methods.

Commit:
Make members private and move a class.

Commit:
Add seat assignment command handling tests.

Commit:
Remove a method, check events order in tests.

Commit:
Finish seat assignments aggregate tests and fixes.

Commit:
Reduce visibility of elements.

Commit:
Add comments.

Commit:
Add seat availability repository and tests.

Commit:
Repository tests refactoring.

Commit:
Remove redundant code.

Commit:
Fix a typo.

Commit:
Update Spine dependency.

Commit:
Improve docs.

Commit:
Improve process manager messages/packages names.

Commit:
Add time to event context to store aggregates successfully.

Commit:
Improve docs.

Commit:
Add registration process manager repository (and a draft of the PM class).

Commit:
Suppress a warning.

Commit:
Update Spine dependencies.

Commit:
Update Spine dependencies.

Commit:
Add todos.

Commit:
Implement RegistrationProcessManager (maybe needs some more changes/refactoring).

Commit:
Change Spine dependency to maven central.

Commit:
Add Travis CI config.

Commit:
Fix build files.

Commit:
Apply proto lookup plugin in modules.

Commit:
Add jacoco plugin.

Commit:
Remove redundant plugin application.

Commit:
Extract a method.

Commit:
Improve logging and fix a context param type.

Commit:
Add PM repo tests.
Fix PM command handler.

Commit:
Rename test method, extract a method and variables.

Commit:
Add some process manager tests.

Commit:
Fix docs.

Commit:
Add more PM tests and fix an error.

AlexanderLitus commented:
@alexander-yevsyukov, PTAL (proc manager is covered 50% now).

Commit:
Fix expiration command handling, etc.

Commit:
Add more PM tests.

Commit:
Extract a method.

Commit:
Add docs.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/procman/RegistrationProcessManagerRepositoryShould.java:
```diff
@@ -0,0 +1,83 @@
+ * @author Alexander Litus
+ */
+@SuppressWarnings("InstanceMethodNamingConvention")
+public class RegistrationProcessManagerRepositoryShould {
```
Do we expect the end user write tests like this?

AlexanderLitus replied:
Why not? I've found several bugs due to these tests.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java:
```diff
@@ -0,0 +1,270 @@
+            return RegistrationProcess.getDefaultInstance();
+        }
+
+        // Is overridden to make accessible in tests.
```
`@VisibleForTesting` annotation also applies here.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java:
```diff
@@ -0,0 +1,270 @@
+        return builder.build();
+    }
+
+    /*package*/ static Timestamp reservationExpirationTimeBeforeNow() {
```
Do we have spaces around `package` in our coding convention? This form is more compact and looks better. If it's not standard we need to update our standard and main code to reflect it. Please take care of it.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java:
```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    /*package*/ static class Event {
```
What do we have this class for? Its name clashes with the class from the main API. Is it to be able to write something like: `Given.Event.orderPlaced()`? If so, please document this.

AlexanderLitus replied:
Yes, but not only for better readability of its usages - to structurize the code also.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java:
```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    /*package*/ static class Command {
```
Is this class named such to be able to write `Given.Command.expireRegistrationProcess()`? If so, please document this.

Commit:
Add @VisibleForTesting annotations.

Commit:
Add useful docs.

Commit:
Format code.

Commit:
Move a class to upper level.

Commit:
Update Spine protobuf plugin dependency.

Commit:
Changes to build.gradle.

Commit:
Update Spine dependency:
remove state neutral events (apply them, add fields to Order);
send ExpireRegistrationProcess command in PM;
add generic param to aggregates;
remove redundant repository tests.

alexander-yevsyukov commented on registration/src/main/java/org/spine3/samples/lobby/registration/conference/ConferenceProjection.java:
```diff
@@ -161,7 +161,7 @@ private void addOrRemoveSeats(int newQuantity, int oldQuantity, SeatTypeId id) {
     }
 
-    // TODO:2016-02-23:alexander.litus: why is it here and not in PM?!
+    // TODO:2016-02-23:alexander.litus: consider moving this method to Process Manager
```
Can we decide on this matter in this PR?

alexander-yevsyukov commented on registration/src/main/java/org/spine3/samples/lobby/registration/conference/ConferenceProjection.java:
```diff
@@ -171,6 +171,7 @@ private void sendAddSeatsRequest(SeatTypeId seatTypeId, int quantity) {
         boundedContext.process(command);
     }
 
+    // TODO:2016-02-23:alexander.litus: consider moving this method to Process Manager
```
Can we decide on this matter in this PR?

alexander-yevsyukov commented on registration/src/main/java/org/spine3/samples/lobby/registration/order/OrderAggregate.java:
```diff
@@ -203,6 +197,27 @@ private void apply(OrderReservationCompleted event) {
+    private void apply(OrderTotalsCalculated event) {
+        final Order.Builder state = getState().toBuilder();
+        state.setPrice(event.getTotal());
+        incrementState(state.build());
```
Aggregates already do not need to call `incrementState()`. Can you migrate this code to latest state of `core-java`?

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/conference/ConferenceProjectionRepositoryShould.java:
```diff
@@ -71,7 +72,7 @@ private TestConferenceProjection(ConferenceId id) {
         }
 
-        // Is overridden to make it accessible in tests.
+        @VisibleForTesting
```
Can we have it package access for tests? Are these tests still required assuming that calling `incrementState()` is no longer needed?

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/order/Given.java:
```diff
@@ -197,36 +206,56 @@ private Command() {
+        /* package */ static OrderRegistrantAssigned orderRegistrantAssigned() {
+            final OrderRegistrantAssigned.Builder builder = OrderRegistrantAssigned.newBuilder()
+                    .setOrderId(ORDER_ID)
+                    .setPersonalInfo(newPersonalInfo("Albert", "Einstein", "ein@stein.com"));
```
Put a email address with a domain of the establishment where Albert last worked.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/seat/assignment/TestSeatAssignmentsAggregate.java:
```diff
@@ -0,0 +1,85 @@
+
+import static com.google.common.base.Throwables.propagate;
+
+/**
```
Please document this class. Its purpose is not clear.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/seat/assignment/TestSeatAssignmentsAggregate.java:
```diff
@@ -0,0 +1,85 @@
+        super(id);
+    }
+
+    // Is overridden to do not throw exceptions while retrieving the default state via reflection.
```
Perhaps we need to fix this matter at the higher level?

Commit:
Update Spine dependency: remove cmd scheduler param.

Commit:
Update Spine dependency: use getBuilder() in Order aggregate event appliers, and new test API.

Commit:
Update Spine dependency: use getBuilder() in Seats Availability aggregate event appliers, and new test API.

Commit:
Update Spine dependency: use getBuilder() in Seats Assignments aggregate event appliers, and new test API.

Commit:
Remove redundant tests.

Commit:
Remove repository tests.

Commit:
Fix Einstein's email.

Commit:
Leave commands sending in the conference projection because this process is tightly coupled with it.

AlexanderLitus commented:
PTAL.

alexander-yevsyukov commented on registration/src/test/java/org/spine3/samples/lobby/registration/seat/availability/testcase/ExistAvailableSeatsAndPendingReservations.java:
```diff
@@ -53,13 +53,13 @@
                 .putAllPendingReservations(PENDING_RESERVATIONS)
                 .build();
-        aggregate.incrementState(state);
+        aggregate.testIncrementState(state);
```
Please rename 
- `testIncrementState()` to `incrementStateForTest()`;
- `testDispatch()` to `dispatchForTest()`.

This way they 1) would sort better with the production counterparts; 2) won't be confused with testing methods.

alexander-yevsyukov replied:
Please also add this convention to our coding guidelines.

Commit:
Update Spine dependency (methods renaming); remove mavenLocal() to avoid stale dependencies.

AlexanderLitus commented:
Done, can I merge it now?

Commit:
Update Spine dependency: Command bus changes, moved classes.

Commit:
Update Spine dependency: 0.3 version; update protobuf-plugin dep-cy.

Commit:
Add comments.

alexander-yevsyukov commented:
LGTM
